Resolve "Refactor Web FTP Command"

geunwoonoh opened the pull request:
As mentioned in https://github.com/alpacanetworks/alpamon-go/issues/3, we have improved the performance and reduced the error rate of `Web FTP command` by implementing the following changes:

- Refactor `Web FTP Command` to utilize Go's built-in packages.
   - To address the performance issue caused by using the `os/exec` package to execute Linux commands, the logic has been refactored to leverage Go's built-in packages.
   
- Add `ftp` alpamon command
   - Given the inefficiency of creating a new, less privileged process for each command using `os/exec`, we have decided to execute `Web FTP Command` in a separate process.
      - To achieve this, we have added the `ftp` command to alpamon, allowing `Web FTP Command` to be executed in a new, less privileged process.
      
- Add `zip` package dependency
   - As mentioned in https://github.com/alpacanetworks/alpamon-go/issues/4, `zip` package dependency has been added to `.goreleaser.yaml` file.

geunwoonoh committed:
Add ftp alpamon command

To execute FtpClient in a process with lowered
privileges, add ftp command to alpamon command.
It accepts url and home directory as arguments to execute FtpClient. The process loads a config to be used.

geunwoonoh committed:
Refactor Web FTP Command

Due to performance issues, refactor the code from
using the os/exec package to using Go's built-in packages.
As a result of changes made to other code to allow FtpClient's RunFtpBackground() to execute in a process with lowered
privileges, this code was also updated.

geunwoonoh committed:
Add FtpSettings

Add FtpSettings and InitFtpSettings
to utilize web FTP in a new process.

geunwoonoh committed:
Update openftp command

Update openftp to run FtpClient in a new process.

geunwoonoh committed:
Add logger in ftp command

Add logger to unify log formats.

geunwoonoh committed:
Minor fix

Add exception handling.

geunwoonoh committed:
Minor fix

Delete unused code
Commit for test.

geunwoonoh committed:
Minor fix

Delete comment.
Commit for test.

geunwoonoh committed:
Fix ftp alpamon command

Update openftp to run FtpClient in a new process with demote permission.

geunwoonoh committed:
Add ModTime at CommandResult

ModTime has been added to the response
of list command in order to display the last
modified time.

geunwoonoh committed:
Add dependencies at .goreleaser.yaml

Add zip package to dependencies so that
it will be installed along with alpamon package.

geunwoonoh committed:
Add InitFtpLogger()

To ensure consistent logging format
with the existing logs, InitFtpLogger()
was added to a new process executing
Web FTP commands.
As the new process has reduced permissions on
/var/alpamon/alpamon.log, InitLogger alone was
insufficient to unify the log structure.
To further consolidate logs from the new process
into the existing log file, Unix domain sockets
should be implemented in the future.

geunwoonoh committed:
Merge remote-tracking branch 'origin/main' into 3-refactor-web-ftp-command

royroyee commented on pkg/runner/ftp.go:
```diff
@@ -401,15 +369,85 @@ func (fc *FtpClient) cpDir(src, dst string) (CommandResult, error) {
+	if err != nil {
+		return err
+	}
+	defer srcFile.Close()
```
Ensure errors are handled explicitly when using defer (Go convention). Review and fix other areas as needed

royroyee commented on pkg/runner/command.go:
```diff
@@ -153,8 +153,35 @@ func (cr *CommandRunner) handleInternalCmd() (int, string) {
 
-		ftpClient := NewFtpClient(cr.data)
-		go ftpClient.RunFtpBackground()
+		sysProcAttr, err := demote(data.Username, data.Groupname)
```
Too much logic in the "openftp" case block. Recommend refactoring into separate functions for readability. Also, avoid reassigning variables like `err` multiple times in the same block.

royroyee commented on pkg/logger/logger.go:
```diff
@@ -64,6 +65,29 @@ func InitLogger() *os.File {
 }
 
+// TODO : Send logs to alpamon's Logserver using a Unix domain socket
+func InitFtpLogger() {
```
This log uses console writer, outputting only to console. Consider following existing logger setup, adjusting for dev and production environments. 
Also, using `log.Logger` as a global variable could lead to clutter.

royroyee replied:
Will different processes started with ftp log to the same file? Not sure if this is the best approach for concurrent logging.

royroyee commented on cmd/alpamon/command/ftp.go:
```diff
@@ -0,0 +1,32 @@
+	"github.com/spf13/cobra"
+)
+
+var ftpCmd = &cobra.Command{
```
`RunE` is used without returning errors. Either add error handling at each step or switch to `Run`.

royroyee commented on pkg/runner/ftp.go:
```diff
@@ -401,15 +369,85 @@ func (fc *FtpClient) cpDir(src, dst string) (CommandResult, error) {
+		return err
+	}
+
+	if err = dstFile.Close(); err != nil {
```
Redundant `dstFile.Close()`

royroyee commented on pkg/runner/ftp.go:
```diff
@@ -401,15 +369,85 @@ func (fc *FtpClient) cpDir(src, dst string) (CommandResult, error) {
+}
+
+func copyDir(src, dst string) error {
+	if strings.HasPrefix(dst, src) {
```
Should the two paths (src and dst) be aligned to the same format? They could differ as relative vs absolute paths – is this accounted for?

geunwoonoh replied:
To support relative paths, `parsePath()` parses relative paths into absolute paths based on the home directory. 
`copyDir()` and other `Web FTP Commands` utilize these parsed paths.

royroyee commented on pkg/runner/command.go:
```diff
@@ -153,8 +153,35 @@ func (cr *CommandRunner) handleInternalCmd() (int, string) {
+		if err := cmd.Start(); err != nil {
+			log.Debug().Err(err).Msg("Failed to start worker process")
+
+			return 1, fmt.Sprintf("openftp: Failed to start worker process. %s", err.Error())
```
Use `err` directly in `fmt.Sprintf` instead of `err.Error()`.

royroyee commented on pkg/runner/command.go:
```diff
@@ -153,8 +153,35 @@ func (cr *CommandRunner) handleInternalCmd() (int, string) {
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+
+		if err := cmd.Start(); err != nil {
```
Please review if it's necessary to manage the child process with `Wait()` here.

geunwoonoh replied:
The purpose of `openftp` is to create a child process, but adding wait() would prevent the command from finishing until the child process ends. This seems inappropriate because the `command fin request` to `alpacon-server` would be sent after the child process terminates.
Instead, I added `os.Exit(1)` to the FtpClient.close() to terminate child process upon completion of the FtpClient connection.

royroyee commented on pkg/config/config.go:
```diff
@@ -18,6 +19,7 @@ var (
 	}
 
 	GlobalSettings Settings
+	FtpSettings    Settings
```
Since ftp runs as a separate process, consider removing global variable for init.

geunwoonoh committed:
Add FtpLogger

Add custom FtpLogger component
to handle logging for isolated FTP processes.

geunwoonoh committed:
Refactor openftp command

To improve readability, encapsulate the logic
of spawning a new, less privileged process into
openftp(), which has been separated from
openftp command.

geunwoonoh committed:
Minor fix

Apply Go coding conventions.
Integrate FtpLogger.

geunwoonoh committed:
Delete FtpSettings

With the FTP process now separated, remove
FtpSettings and have NewFtpClient fetch
settings directly.

geunwoonoh committed:
Update ftp command due to refactoring

Update related to the addition of FtpLogger
and changes to how settings are retrieved.

geunwoonoh committed:
Add os.Exit(1) to manage child process

To ensure that child process terminates
upon the completion of the FtpClient connection,
add os.Exit(1).

geunwoonoh committed:
Fix config loading error

Permissions error prevented new process from
loading the config file when run NewFtpClinet().
To resolve this, ftp command was refactored
to load the config and pass the configuration
as FtpConfigData to the newly spawned
NewFtpClient process.

geunwoonoh committed:
Minor fix

Apply golang convention.

royroyee committed:
Minor fix

geunwoonoh committed:
Minor fix

Set the initial value of finalErr to nil.
Fix error handling within the defer() function
to populate finalErr with any encountered
errors.

royroyee committed:
Enhance error handling in copyFile function
